python-spec/src/somacore/base.py

```diff
@@ -24,7 +25,7 @@ def open(
         uri: str,
         mode: options.OpenMode = "r",
         *,
-        context: Optional[Any] = None,
+        context: Optional[ContextBase] = None,
```

Comment:
I was thinking about this a little further and it turns out there is a deeper wrinkle to this: we wouldn’t be able to override

```python

class SOMAObject(...):
  @classmethod
  def open(..., context: Optional[ContextBase] = None) -> Self: ...
```

with

```python
class ImplObject(somacore.SOMAObject):
  @classmethod
  def open(..., context: Optional[ImplContext] = None) -> Self: ...
  # not allowed, because `ImplContext` is a narrowing of `ContextBase`,
  # which would violate the Liskov principle:
  # https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

We could do a bunch of chicanery with another entry in the list of generics, so it would look something like:

```python
_CtxT = TypeVar("_CtxT", bound=ContextBase)
class SOMAObject(Generic[_CtxT]):
  def open(..., context: Optional[_CtxT] = None) -> Self: ...

# elsewhere:

class ImplObject(somacore.SOMAObject[ImplContext]):
  def open(..., context: Optional[ImplContext] = None) -> Self: ...
```

but that would also require adding a `_CtxT` to *every* SOMA object, which would be particularly unwieldy on the already highly-genericised Measurement type:

https://github.com/single-cell-data/SOMA/blob/61f6cc9e1608bd4a33110056e0530e434c2e1495/python-spec/src/somacore/measurement.py#L26-L29

so instead, I think the easier thing to do is to opt out of type-checking for this parameter when being passed in, which, while imperfect, will allow it to be overridden with type information in the implementation. So the end result looks like:

```python
class SOMAObject(metaclass=abc.ABCMeta):
  @classmethod
  def open(..., context: Optional[Any] = None) -> Self:
    ...
  # Ignore type-checking the specific type allowed in `open`

  @property
  def context(self) -> Optional[types.ContextBase]: ...
  # Returning a subclass is OK, though.
```

Then, at the implementation, we can say:

```python
class ImplObject(SOMAObject):
  @classmethod
  def open(..., context: Optional[ImplContext] = None) -> Self: ...
  # Allowed since the superclass `context` parameter type is unchecked.

  @property
  def context(self) -> Optional[ImplContext]: ...
  # Allowed since you can narrow returned types.
```